pkg/perf/pubip_task.go

```diff
@@ -0,0 +1,238 @@
+package perf
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/rs/zerolog/log"
+	substrate "github.com/threefoldtech/tfchain/clients/tfchain-client-go"
+	"github.com/threefoldtech/zos/pkg/environment"
+	"github.com/threefoldtech/zos/pkg/network/macvlan"
+	"github.com/threefoldtech/zos/pkg/network/namespace"
+	"github.com/threefoldtech/zos/pkg/stubs"
+	"github.com/vishvananda/netlink"
+)
+
+const testMacvlan = "pubtestmacvlan"
+const testNamespace = "pubtestns"
+
+type publicIPValidationTask struct {
+	taskID   string
+	schedule string
+}
+
+var _ Task = (*publicIPValidationTask)(nil)
+
+func NewpublicIPValidationTask() Task {
+	return &publicIPValidationTask{
+		taskID:   "PublicIPValidation",
+		schedule: "0 0 */6 * * *",
+	}
+}
+
+func (p *publicIPValidationTask) ID() string {
+	return p.taskID
+}
+
+func (p *publicIPValidationTask) Cron() string {
+	return p.schedule
+}
+
+func (p *publicIPValidationTask) Run(ctx context.Context) (interface{}, error) {
+
+	netNS, err := namespace.GetByName(testNamespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get namespace %s: %w", testNamespace, err)
+	}
+	manager, err := environment.GetSubstrate()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get substrate client: %w", err)
+	}
+	sub, err := manager.Substrate()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get substrate client: %w", err)
+	}
+	defer sub.Close()
+	farmID := environment.MustGet().FarmID
+
+	shouldRun, err := isLeastValidNode(ctx, uint32(farmID), sub)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if the node should run public IP verification: %w", err)
+	}
+	if !shouldRun {
+		log.Info().Msg("skipping because there is a node with less ID available")
+		return nil, nil
+	}
+
+	farm, err := sub.GetFarm(uint32(farmID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get farm with id %d: %w", farmID, err)
+	}
+	unusedIPs := make(map[string]bool)
+	err = netNS.Do(func(nn ns.NetNS) error {
+		mv, err := macvlan.GetByName(testMacvlan)
+		if err != nil {
+			return fmt.Errorf("failed to get macvlan %s in namespace %s: %w", testMacvlan, testNamespace, err)
+		}
+		for _, publicIP := range farm.PublicIPs {
+			if publicIP.ContractID != 0 {
+				continue
+			}
+			unusedIPs[publicIP.IP] = false
+
+			ip, ipNet, routes, err := getIPWithRoute(publicIP)
+			if err != nil {
+				log.Err(err).Send()
+				continue
+			}
+			err = macvlan.Install(mv, nil, ipNet, routes, netNS)
```

Comment:
I don't believe you need to pass netNS because u are already inside the namespace. it's then better to pass `nil` to avoid unpredeceted errors